### PR TITLE
Diameter of a binary tree in C++

### PR DESCRIPTION
**Pull request by zhcet19**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

Resolved issue #3056 

(Write your answer here.)

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

(Write your answer here.)

### CODE REVIEW

**Comment by vanshika2906** [comment]
I would like to work on this. Please /assign it to  me.

**Comment by zhcet19** [comment]
@iamrajiv please review my pr

**Comment by zhcet19** [comment]
@iamrajiv please review my pr

**Comment by zhcet19** [comment]
@iamrajiv please review my pr

**Review by iamrajiv — APPROVED** [review]

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  Definition:-The diameter of a binary tree (sometimes called the width) is the number of nodes
+  on the longest path between two end nodes.
+
+   Approach of solving:-Diameter of a tree can be calculated by utilizing the property that
+   the diameter of a tree is nothing but maximum value of (left_height + right_height + 1) for each node.
+   So we need to calculate this value (left_height + right_height + 1) for each node and update the result. 
+   
+*/
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: - Remove extra lines of spaces

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+#include <bits/stdc++.h>
+using namespace std;
+
+//User-define Tree structure
+struct  Tree {
+    int data;
+    Tree* left, *right;
+};
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: - Remove extra one lines of spaces

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    if (node == NULL)
+    {
+         return 0;
+    }
+       
+    //Case 2:- height would be 1+max(height of left tree,height of right tree)
+    return 1 + max(height(node->left), height(node->right));
+}
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: - Remove extra one lines of spaces

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+int Diameter(struct Tree* node)
+{
+    //  case 1:- when Tree is empty
+    if (node == NULL)
+    {
+        return 0;  
+    }
+      
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: - Remove extra two lines of spaces

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        return 0;  
+    }
+      
+
+ 
+    //  Finding the  diameter of left and right sub-trees
+    int leftdiameter = Diameter(node->left);
+    int rightdiameter = Diameter(node->right);
+    
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: - Remove extra one lines of spaces

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    
+    // Finding the height of left and right sub-trees
+    int leftheight = height(node->left);
+    int rightheight = height(node->right);
+ 
+    /* 
+      Result is nothing but the maximum of diameter of left,right  subtrees and 
+      height of left subtree + height of right subtree + 1;
+    
+    */
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: - Remove extra lines of spaces and redundant comments

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    return max(leftheight + rightdiameter + 1, max(leftdiameter, rightdiameter));
+}
+int main()
+{
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: - Remove extra lines of spaces

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    root->left->right = CreateNode(6);
+    root->left->right->left = CreateNode(10);
+    
+    cout<<"Diameter of the above tree is:-"<<Diameter(root)<<"\n";
+   
+    return 0;
+}
+
+
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: - Remove extra two lines of spaces

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

**Comment by zhcet19** [comment]
I have removed extra spaces please review it now @ankitaggarwal23

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+*/
+#include <bits/stdc++.h>
+using namespace std;
+
+//User-define Tree structure
+struct  Tree {
+    int data;
+    Tree* left, *right;
+};
+struct Tree* CreateNode(int data)
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp:22]: Add a new line after this };

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+{
+    struct Tree* node = new Tree;
+    node->data = data;
+    node->left = NULL;
+    node->right = NULL;
+ 
+    return (node);
+}
+int height(struct Tree* node)
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp:32]: Add a new line after this }

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+{
+     //  case 1:- when Tree is empty
+    if (node == NULL)
+    {
+         return 0;
+    }
+    //Case 2:- height would be 1+max(height of left tree,height of right tree)
+    return 1 + max(height(node->left), height(node->right));
+}
+int Diameter(struct Tree* node)
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp:43]: Add a new line above int Diameter

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    // Finding the height of left and right sub-trees
+    int leftheight = height(node->left);
+    int rightheight = height(node->right);
+    /* 
+      Result is nothing but the maximum of diameter of left,right  subtrees and 
+      height of left subtree + height of right subtree + 1;
+    */
+    return max(leftheight + rightdiameter + 1, max(leftdiameter, rightdiameter));
+}
+int main()
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp:64]: Add a new line above the int main()

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+Input Output Analysis:-
+Input Tree:-  1
+            / \
+          4    9
+        / \
+      3   6
+         /
+         10
+      Output:-5
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: Capture

You Code prints the output like this kindly rewrite in a proper way.

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]
Removing extra spaces means if you have more than one line space in your code just remove that extra line.

**Comment by zhcet19** [comment]
I have added changes,kindly review it

File: C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+Input Output Analysis:-
+Input Tree:- 1
+            / \
+           4   9
+          / \  
+         3  6
+           /
+         10
+         
+      Output:-Diameter of the above tree is:-5
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/graphs/Diameter_of_binary_tree.cpp]: Remove that Extra space before Output.
`Output:-` 
`Diameter of the above tree is:-5`

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]
Rest looks good to merge.

**Comment by zhcet19** [comment]
Removed extra spacing before output

**Review by ankitaggarwal23 — APPROVED** [review]

**Comment by zhcet19** [comment]
@HarshCasper please merge this pull request

**Pull request merged**